api/ombpdf/semdb.py

```diff
@@ -1,12 +1,153 @@
+
+    @pdf_node.setter
+    def pdf_node(self, value):
+        self.tree.node[self.identifier]['pdf_node'] = value
 
 
 def to_db(doc, policy):
-    dom = semdom.to_dom(doc)
-    xml = etree.fromstring(dom.toxml())
-    root = import_xml_doc(policy, xml, ignore_preamble=True)
-    return root
+    writer = DatabaseWriter(policy)
+    builder = semtree.SemanticTreeBuilder(doc, writer)
+    builder.build()
+
+    writer.cursor.nested_set_renumber()
+    # Account for a Django bug which prevents us from saving these
+    # directly due to saving the DocNodes in a bulk_insert
+    for footnote_citation in writer.footnote_citations.values():
+        footnote_citation.doc_node = footnote_citation.doc_node
+        footnote_citation.footnote_node = footnote_citation.footnote_node
+        footnote_citation.save()
+
+    return writer.cursor
+
+
+class DatabaseWriter(semtree.Writer):
+    def __init__(self, policy):
+        self.cursor_stack = [PDFAwareCursor.new_tree('policy', policy=policy)]
+        self.footnote_citations = {}
+
+    @property
+    def cursor(self):
+        return self.cursor_stack[-1]
+
+    @property
+    def sec_level(self):
+        return self.cursor.identifier.split('_').count('sec')
+
+    def end_element(self, el):
+        self.cursor.model.text = self.cursor.text.rstrip()
+        super().end_element(el)
+
+    def begin_document(self, doc):
+        pass
+
+    def end_document(self, doc):
+        self.cursor_stack = self.cursor_stack[:1]
+
+    def begin_heading(self, heading):
+        while heading.level <= self.sec_level:
+            self.cursor_stack.pop()
+        while heading.level > self.sec_level:
+            self.cursor_stack.append(self.cursor.add_child('sec'))
+        self.cursor_stack.append(self.cursor.add_child('heading'))
+        self.cursor.pdf_node = heading
+
+    def end_heading(self, heading):
+        text = self.cursor.text.strip()
+        self.cursor_stack.pop()
+        max_length = DocNode._meta.get_field('title').max_length
+        if len(text) > max_length:
+            logger.warning('Heading too long. Misparse? %s', text)
+        self.cursor.model.title = text[:max_length]
+
+    def begin_paragraph(self, p):
+        self.cursor_stack.append(self.cursor.add_child('para'))
+        self.cursor.pdf_node = p
+
+    def end_paragraph(self, p):
+        self.cursor_stack.pop()
+
+    def begin_list(self, li):
+        self.cursor_stack.append(self.cursor.add_child('list'))
+        self.cursor.pdf_node = li
+
+    def end_list(self, li):
+        self.cursor_stack.pop()
+
+    def begin_list_item(self, p):
+        li = self.cursor.add_child('listitem')
+        self.cursor_stack.append(li.add_child('para'))
+        self.cursor.pdf_node = p
+
+        parent_list = next(
+            self.cursor.ancestors(lambda n: n.node_type == 'list'),
+            None,
+        )
+        if not parent_list:
+            logger.warning('Adding list item, but no list parent')
```

Comment:
I remember you mentioning this to me on Slack--is it still a problem?  If so, we should file an issue, because it'd definitely be a bug in `semtree.py`!

Reply:
Indeed it is; created #803